Draft for release procedure. (#2325)
radareorg/cutter: docs/source/contributing/code/release-procedure.rst

8. If any major problems are found fix them. If a release branch is used fix them in master and cherry pick into release branch. If the amount of changes is sufficiently large repeat from step 3. increasing rc number by one.
9. Update version to 1.11.0
10. Create tag
11. Create release
    * Fill the release notes, preparing release notes can be started earlier. Compare current master or release branch against previous release to find all the changes. Choose the most important ones. Don't duplicate the commit log. Release notes should be a summary for people who don't want to read whole commit log. Group related changes.

Basic testing procedure
-----------------------

This isn't intended as exhaustive testing process, just some simple steps to make sure everything isn't badly broken.

* Open simple executable like /bin/ls or calc.exe
* Make sure upgraded layout isn't completely broken
* Disassembly shows something
* Bundled plugins work
   * Open decompiler and select ghidra, it shows some C code at least for some functions
   * Open r2dec in decompiler widget, make sure it shows code
* Test that sample python plugin works
* Try debugger
   * Insert breakpoint in main

>    * Start debuging

Reviewer: debugging*

Release Procedure
=================

1. Update translations submodule
    1. Update the translations in cutter-translations from Crowdin. The latest archive can only be generated by users with appropriate privileges.

>     2. Place the archive in folder and run the script, see cutter-translations for instructions

Reviewer 2: mention its name `build.sh`.
What instructions are available ion the repo?
make it linkable

Author: How about now?

    3. Commit to cutter-translations
    4. Update submodule in cutter

> 2. If there is desire to keep working in the master, create branch for the release and do all the following work there.

Reviewer 2: ```suggestion
2. If there is a desire to keep working in the master branch, create branch for the release and do all the following work there.
```

> 3. Lock r2ghidra and r2dec versions downloaded by packaging scripts. Specify a tag or commit hash.

Reviewer 2: how? where?

> 4. Update version

Reviewer 2: ```suggestion
4. Update the version of Cutter by bumping the strings and configurations that are stating the version
```

    #. appveyor.yml
    #. docs/sourc/conf.py
    #. docs/source/index.rst
    #. Cutter.pro
    #. Cutter.appdata.xml

>     #. To be safe search for the previous version number.

Reviewer 2: ```suggestion
    #. To be safe, search the code base for the previous version number.
```

5. Create a tag v1.11.0-rc1
    #. ``git tag v1.11.0-rc1``

>     #. ``git tag push origin v1.11.0-rc1``

Reviewer 2: Why specific to v1.11?

Author: It seemed easier to explain using a specific examples.

Reviewer 2: I think it is more confusing. For example, here it's harder to imaging patch-version upgrade as the patch-version is `0`. Best solution in my opinion is to give give several examples as this can only be 

Current version: `x.y.z`
New tag: `x.y.[z+1]-rc1` or `x.[y+1].0-rc1` or `[x+1].0.0-rc1`

What I suggest is something like this:
```
5. Create a tag for the release candidate. For example, for the `v1.11.0` release you'd do something like this:
    #. ``git tag v1.11.0-rc1``
    #. ``git tag push origin v1.11.0-rc1``
```

Author: Feel free to modify the parts referring to version in the way you think is better.

> 6. Create a GitHub release, mark it as pre-release save draft, set the tag to v1.11.0-rc1

Reviewer 2: Specific tag is used

> 7. Wait for packages to build

Reviewer 2: "Because we already pushed the new tag, GitHub knows to automatically start a building process for the packages and place them as artifacts in the release. Wait for the packages to be built."

8. On all operating systems do the `Basic testing procedure`_ to ensure nothing badly broken.

> 9. If any major problems are found fix them. If a release branch is used fix them in master and cherry pick into release branch. If the amount of changes is sufficiently large repeat from step 3. increasing rc number by one.

Reviewer 2: ```suggestion
9. If any major problems are found, open an issue and fix them. If a release branch is used fix them in master and cherry pick into release branch. If the amount of changes is sufficiently large repeat from step 3. increasing rc number by one.
```

> 10. Update version to 1.11.0

Reviewer 2: specific version

Reviewer 2: remind them to remove the release candidate part from the version

> 11. Create tag

Reviewer 2: Duplicate the code example you showed before (for `rc`), it is useful

12. Create release

>     * Fill the release notes, preparing release notes can be started earlier. Compare current master or release branch against previous release to find all the changes. Choose the most important ones. Don't duplicate the commit log. Release notes should be a summary for people who don't want to read whole commit log. Group related changes.

Reviewer 2: ```suggestion
    * Fill the release notes in the Release description. Preparing release notes can begin earlier. Compare current master or release branch against previous release to find all the changes. Choose the most important ones. Don't duplicate the commit log. Release notes should be a summary for people who don't want to read whole commit log. Group related changes together under titles such as "New features", "Bug Fixes", "Decompiler", "radare2" and similar.
```

> 13. Announcement

Reviewer 2: ```suggestion
13. Prepare announcement tweets and messages to send in the Telegram group, reddit, and others.
```

> 14. Close milestone if there was one

Reviewer 2: Maybe open a new one as well?

Bugfix Release
--------------
The process for bugfix release is similar no normal release procedure described above with some differences.

* Create the branch from the last tag instead of master or reuse the branch from x.y.0 release if it was already created.
* Cherry pick required bugfixes from master into the branch.
* Increase the third version number x.y.n into x.y.(n+1) .

Basic testing procedure
-----------------------

> This isn't intended as exhaustive testing process, just some simple steps to make sure everything isn't badly broken.

Reviewer 2: ```suggestion
This isn't intended as exhaustive testing process, just some simple steps to make sure nothing is badly broken.
```

Author: You have higher requirements :smile:

> * Open simple executable like /bin/ls or calc.exe

Reviewer 2: ```suggestion
* Open a simple executable like ``/bin/ls`` or ``calc.exe``
```

> * Make sure upgraded layout isn't completely broken

Reviewer 2: ```suggestion
* Make sure that the upgraded layout isn't completely broken
```

> * Disassembly shows something

Reviewer 2: ```suggestion
* The Disassembly widget shows proper disassembly.
```

* Bundled plugins work
   * Open decompiler and select ghidra, it shows some C code at least for some functions

>    * Open r2dec in decompiler widget, make sure it shows code

Reviewer 2: Check multiple functions and see that they are not causing a crash.

we had the situation in which r2dec crashed for some functions

Author: That could be a general note for any of the steps in testing procedure. Try a few times with different functions or offsets.|

Author: Added

> * Test that sample python plugin works

Reviewer 2: what about sample Cpp plugin?

Author: Currently  cpp plugins can't be properly built against release package after it has been built. After some of the CMake refactoring is finished it should be more feasible.

Release Procedure
=================

1. Update translations submodule `https://github.com/radareorg/cutter-translations`_
    1. Get the latest archive from Crowdin. The latest archive can only be generated by users with appropriate privileges.
    2. Place the archive in folder and run the build.sh script, see `cutter-translations <https://github.com/radareorg/cutter-translations>`_ for any changes in process
    3. Commit to cutter-translations (only the ts files)
    4. Update submodule in cutter
2. If there is a desire to keep working in the master branch, create branch for the release and do all the following work there.
3. Lock r2ghidra and r2dec versions downloaded by packaging scripts. Specify a tag or commit hash.
4. Update version
    #. appveyor.yml
    #. docs/sourc/conf.py
    #. docs/source/index.rst
    #. Cutter.pro
    #. Cutter.appdata.xml
    #. To be safe, search the code base for the previous version number.

> 5. Create a tag v1.11.0-rc1
>     #. ``git tag v1.11.0-rc1``
>     #. ``git tag push origin v1.11.0-rc1``

Reviewer 2: ```suggestion
5. Create a tag for the release candidate. For example, for the `v1.11.0` release you'd do something like this:
    #. ``git tag v1.11.0-rc1``
    #. ``git tag push origin v1.11.0-rc1``
```
